euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/windowing/WindowOperator.java

```diff
@@ -0,0 +1,465 @@
+
+    this.timerService =
+            getInternalTimerService("window-timers", windowSerializer, this);
+    this.endOfStreamTimerService =
+            getInternalTimerService("end-of-stream-timers", windowSerializer, this);
+    this.triggerContext = new TriggerContextAdapter();
+    this.outputContext = new OutputContext();
+    this.storageProvider = new WindowedStorageProvider<>(
+            getKeyedStateBackend(), windowSerializer);
+
+    this.windowStates = HashBasedTable.create();
+
+    if (windowing instanceof MergingWindowing) {
+      TupleSerializer<Tuple2<WID, WID>> tupleSerializer =
+              new TupleSerializer<>((Class) Tuple2.class, new TypeSerializer[]{windowSerializer, windowSerializer});
+      this.mergingWindowsDescriptor = new ListStateDescriptor<>("merging-window-set", tupleSerializer);
+    }
+  }
+
+  private void setupEnvironment(Object key, WID window) {
+    outputContext.setKey(key);
+    outputContext.setWindow(window);
+    storageProvider.setWindow(window);
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void processElement(StreamRecord<KeyedMultiWindowedElement<WID, ?, ?>> record)
+          throws Exception {
+
+    KeyedMultiWindowedElement<WID, ?, ?> element = record.getValue();
+
+    // drop late-comers immediately
+    if (element.getTimestamp() < timerService.currentWatermark()) {
+      return;
+    }
+
+    if (windowing instanceof MergingWindowing) {
+      MergingWindowSet<WID> mergingWindowSet = getMergingWindowSet();
+
+      for (WID window : element.getWindows()) {
+
+        // when window is added to the set it may result in window merging
+        WID currentWindow = mergingWindowSet.addWindow(window,
+                (mergeResult, mergedWindows, stateResultWindow, mergedStateWindows) -> {
+
+                  setupEnvironment(getCurrentKey(), mergeResult);
+                  triggerContext.setWindow(mergeResult);
+
+                  processTriggerResult(mergeResult,
+                          triggerContext.onMerge(mergedWindows),
+                          mergingWindowSet);
+
+                  // clear all merged windows
+                  for (WID merged : mergedWindows) {
+                    storageProvider.setWindow(merged);
+                    trigger.onClear(merged, triggerContext);
+                    removeWindow(merged, null);
+                  }
+
+                  // FIXME This implementation relies on fact that
+                  // stateCombiner function is actually "fold left".
+                  // That means the result state will end up in the
+                  // first state given to the combiner function.
+
+                  // merge all mergedStateWindows into stateResultWindow
+                  List<State> states = new ArrayList<>();
+                  states.add(getWindowState(stateResultWindow));
+                  mergedStateWindows.forEach(sw -> states.add(getWindowState(sw)));
+                  stateCombiner.apply(states);
```

Comment:
in regards to the above FIXME, i wanted to suggest to change the type of the state-combiner. however, now i see it's a `CombinableReduceFunction` which does have a return value that is supposed to replace the merged states. in the `inmem` as well as in the spark executor (e.g. [GroupedReducer](https://github.com/seznam/euphoria/blob/master/euphoria-core/src/main/java/cz/seznam/euphoria/core/executor/greduce/GroupReducer.java#L182)) we are doing so. if i'm not mistaken applying the same technique here should resolve the above FIXME.